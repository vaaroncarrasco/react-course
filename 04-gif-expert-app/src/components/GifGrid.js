import React from 'react'
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    // Every time theres a change in component, it runs the whole code all over again and again -> useEffect fixes this by calling functions once

    const { data:images, loading } = useFetchGifs( category );

    return (
        <>
            <h3> { category } </h3>

            { loading && <p>Loading...</p> }

            <div className="card-grid">
                {
                    images.map(img => (
                        <GifGridItem
                            key={ img.id }
                            // * spreading img{} props and sending em all at once instead of prop={ prop }
                            { ...img }
                            // img={ img }
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}