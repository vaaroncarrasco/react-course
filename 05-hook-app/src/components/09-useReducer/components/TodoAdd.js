import React from 'react'
import { useForm } from '../../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ { description }, handleInputChange, reset ] = useForm({ // useForm returns [values{}, handleInputChange]
        // same as name="description" in html attributes
        description: '',
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if ( description.trim().length <= 1) return;

        // * new object
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        // * dispatch sends action to reducer
        handleAddTodo( newTodo );
        reset();

    }

    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr />

            <form onSubmit={ handleSubmit }>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>

            </form>
        </>
    )
}
