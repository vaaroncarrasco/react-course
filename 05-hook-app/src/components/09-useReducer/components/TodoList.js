import React from 'react'
import { TodoListItem } from './TodoListItem'
import PropTypes from 'prop-types';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map( (todo, index) => ( // * () => () -> () implicitly returning an object
                    <TodoListItem
                        key={ todo.id }
                        todo={ todo }
                        index={ index }
                        handleDelete={ handleDelete }
                        handleToggle={ handleToggle }
                    />
                ))
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}
