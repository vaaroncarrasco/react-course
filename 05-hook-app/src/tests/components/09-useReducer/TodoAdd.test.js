import React from 'react';
import { shallow } from 'enzyme';
import { TodoAdd } from '../../../components/09-useReducer/components/TodoAdd';

describe('testing <TodoAdd />', () => {

    const handleAddTodo = jest.fn();

    const wrapper = shallow(
        <TodoAdd
            handleAddTodo={ handleAddTodo }
        />
    );

    test('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should NOT call handleAddTodo', () => {

        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({ preventDefault(){} }) // call it sending (e) -> e.preventDefault()

        expect( handleAddTodo ).toHaveBeenCalledTimes(0);
    });

    test('should call handleAddTodo', () => {
        // con un argumento/called
        // simulate input change // submit form // fn to have been called

        const value = 'Aprender MERN';
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({ preventDefault(){} }) // call it sending (e) -> e.preventDefault()

        expect( handleAddTodo ).toHaveBeenCalledTimes(1);
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) );
        expect( handleAddTodo ).toHaveBeenCalledWith({
            id: expect.any(Number), // ? timestamp id check
            desc: value,
            done: false
        });

        expect( wrapper.find('input').prop('value') ).toBe('');

    });

});