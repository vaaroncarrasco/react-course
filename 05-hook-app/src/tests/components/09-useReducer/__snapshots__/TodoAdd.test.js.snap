// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <TodoAdd /> should match snapshot 1`] = `
<Fragment>
  <h4>
    Agregar ToDo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
